File: server_client_java_protoV3/JavaSocketsFestival_proto/Java11ChatSockets/ChatServer/src/main/java/festival/server/Service.java
package festival.server;


import festival.model.*;
import festival.persistence.repository.DBRepoAngajat;
import festival.persistence.repository.Repository;
import festival.services.ChatException;
import festival.services.IChatObserver;
import festival.services.IChatServices;


import java.time.LocalDateTime;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class Service  <ID, E extends Identifier<ID>> implements IChatServices {
    private Repository<Long, Bilet> repository_bilet;
    private Repository<Long, Artist> repository_artist;
    private Repository<Long, Spectecol> repository_spectacol;
    private Repository<Long, Angajat> repository_angajat;
    private Repository<Long, Client> repository_client;
    private Repository<Long, Persoana> repository_persoana;


    private Map<Long, IChatObserver> loggedClients;

    public Service(Repository<Long, Bilet> repo_b,
                   Repository<Long, Artist> repo_artist,
                   Repository<Long, Spectecol> repo_spectacol,
                   Repository<Long, Angajat> repo_angajat,
                   Repository<Long, Client> repo_client,
                   Repository<Long, Persoana> repo_persoana) {
        this.repository_bilet = repo_b;
        this.repository_artist = repo_artist;
        this.repository_spectacol = repo_spectacol;
        this.repository_angajat = repo_angajat;
        this.repository_client = repo_client;
        this.repository_persoana = repo_persoana;
        loggedClients = new ConcurrentHashMap<>();

    }

    @Override
    public List<String> findArtistiSpectacoleBilete() {

        List<String> listaInregistrariBilete = new ArrayList<>();

        for (Artist artist : repository_artist.findAll()) {
            for (Spectecol spectacol : repository_spectacol.findAll()) {
                if(spectacol.getId_artist()==artist.getId())
                {
                    int nr_bilete_date=0;
                    for (Bilet bilet : repository_bilet.findAll()) {
                        if(bilet.getId_spectacol()==spectacol.getId())
                        {
                            nr_bilete_date+=bilet.getNr_locuri();
                        }
                    }
                    listaInregistrariBilete.add(artist.getNume()+' '+artist.getPrenume()+','
                            + spectacol.getData_inc()+','+spectacol.getLocatie()+','+spectacol.getNr_locuri()+','+nr_bilete_date);
                }
            }
        }


        return listaInregistrariBilete;
    }

    @Override
    public List<String> findArtistiSpectacoleBileteDupaData(LocalDateTime date) {
        List<String> listaInregistrariBilete = new ArrayList<>();

        for (Artist artist : repository_artist.findAll()) {
            for (Spectecol spectacol : repository_spectacol.findAll()) {
                if(spectacol.getData_inc().getDayOfMonth()==date.getDayOfMonth()&&spectacol.getData_inc().getMonth()==date.getMonth()&&spectacol.getData_inc().getYear()==date.getYear())
                    if(spectacol.getId_artist()==artist.getId())
                    {
                        int nr_bilete_date=0;
                        for (Bilet bilet : repository_bilet.findAll()) {
                            if(bilet.getId_spectacol()==spectacol.getId())
                            {
                                nr_bilete_date+=bilet.getNr_locuri();
                            }
                        }
                        listaInregistrariBilete.add(artist.getNume()+' '+artist.getPrenume()+','
                                + spectacol.getData_inc()+','+spectacol.getLocatie()+','+spectacol.getNr_locuri()+','+nr_bilete_date);
                    }
            }
        }
        return listaInregistrariBilete;
    }

    @Override
    public Spectecol findArtistiSpectacoleBileteIndex(int index) {
        List<String> listaInregistrariBilete = new ArrayList<>();
        int i = 0;
        for (Artist artist : repository_artist.findAll()) {
            for (Spectecol spectacol : repository_spectacol.findAll()) {
                if (spectacol.getId_artist() == artist.getId()) {
//                    int nr_bilete_date = 0;
//                    for (Bilet bilet : repository_bilet.findAll()) {
//                        if (bilet.getId_spectacol() == spectacol.getId()) {
//                            nr_bilete_date += bilet.getNr_locuri();
//                        }
//                    }
                    if (i == index)
                        return spectacol;
//                    listaInregistrariBilete.add(artist.getNume() + ' ' + artist.getPrenume() + ','
//                            + spectacol.getData_inc() + ',' + spectacol.getLocatie() + ',' + spectacol.getNr_locuri() + ',' + nr_bilete_date);
                    i += 1;
                }
            }
        }
        return null;
    }

    @Override
    public Spectecol findArtistiSpectacoleBileteDupaDataIndex(String text ) {
        String[] parti = text.split("\\s+");

// Extrage prima parte (reprezentarea LocalDateTime)
        String dateAsString = parti[0];

// Extrage a doua parte (indexul)
        int index = Integer.parseInt(parti[1]);

// Convertirea reprezentării LocalDateTime la obiect LocalDateTime
        LocalDateTime date = LocalDateTime.parse(dateAsString);

        List<String> listaInregistrariBilete = new ArrayList<>();
        int i=0;
        for (Artist artist : repository_artist.findAll()) {
            for (Spectecol spectacol : repository_spectacol.findAll()) {
                if(spectacol.getData_inc().getDayOfMonth()==date.getDayOfMonth()&&spectacol.getData_inc().getMonth()==date.getMonth()&&spectacol.getData_inc().getYear()==date.getYear())
                    if(spectacol.getId_artist()==artist.getId())
                    {
//                        int nr_bilete_date=0;
//                        for (Bilet bilet : repository_bilet.findAll()) {
//                            if(bilet.getId_spectacol()==spectacol.getId())
//                            {
//                                nr_bilete_date+=bilet.getNr_locuri();
//                            }
//                        }
                        if(i==index)
                            return spectacol;
                        i+=1;
//                        listaInregistrariBilete.add(artist.getNume()+' '+artist.getPrenume()+','
//                                + spectacol.getData_inc()+','+spectacol.getLocatie()+','+spectacol.getNr_locuri()+','+nr_bilete_date);
                    }
            }
        }
        return null;
    }

    @Override
    public Long savePersoanaReturnIdPers(String nume, String prenume) {
        Persoana persoana=new Persoana(nume,prenume);
        repository_persoana.save(persoana);
        List <Persoana> p= (List<Persoana>) repository_persoana.findAll();

        return p.get(p.size() - 1).getId();
    }

    @Override
    public Long saveClientReturnIdClient(String nume, String prenume) {
        Long id=savePersoanaReturnIdPers(nume,prenume);
        Client pers=new Client(id,nume,prenume);
        repository_client.save( pers);
        List <Client> p= (List<Client>) repository_client.findAll();

        return p.get(p.size() - 1).getId();
    }
    private final int defaultThreadsNo=5;

    @Override
    public Bilet saveBilet(String text) {
        System.out.println(text);
        String[] parti = text.split("\\s+"); // "\\s+" înseamnă "orice caracter de spațiu, una sau mai multe apariții"

// Atribuirea fiecărei părți la variabilele corespunzătoare
        String nume = parti[0];
        String prenume = parti[1];
        int numarLocuriB = Integer.parseInt(parti[2]);
        Long IDAngajat = Long.parseLong(parti[3]);
        Long IDSpectacol = Long.parseLong(parti[4]);

        Long idClient=saveClientReturnIdClient(nume,prenume);
        Bilet bilet=new Bilet(IDAngajat,idClient,IDSpectacol,numarLocuriB);
        repository_bilet.save(bilet);

        ExecutorService executor= Executors.newFixedThreadPool(defaultThreadsNo);
        for(IChatObserver observer : loggedClients.values()){
            try{
                observer.rezervareReceived(bilet);
            }catch (ChatException e){
                System.out.println("Eror: " + e.getMessage());
            }
        }

        executor.shutdown();
        System.out.println(bilet.getId());

        return bilet;
    }

    public synchronized Angajat valid(String pass, String u, IChatObserver observer) throws ChatException{
        // Apelăm metoda authenticate
        System.out.println(" login, service");
        Angajat userFound =  ((DBRepoAngajat) repository_angajat).authenticate(u, pass);
        if(userFound != null){
            if(loggedClients.get(userFound.getId())!=null)
                throw new ChatException("User already logged in. ");
            loggedClients.put(userFound.getId(), observer);
            return userFound;
        }else{
//            throw new ServerException("Authentication failed. ");
            return null;

        }
    }

    public synchronized void logout(Angajat user, IChatObserver observer) throws ChatException {
//        , IObserver client
        Angajat userFound = ((DBRepoAngajat) repository_angajat).authenticate(user.getUsername(), user.getPassword());
        IChatObserver localClient = loggedClients.remove(userFound.getId());
        if(localClient == null)
            throw new ChatException("User " + userFound.getId() + " is not logged in. ");
    }

}

File: server_client_C_sharp/mpp_c_app1/FestivalServer/service.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using FestivalModel;
using FestivalPersistence;
using FestivalService;

namespace FestivalServer
{
    public class Service:IService
    {

        private IRepository<long, Bilet> repository_bilet;
        private IRepository<long, Artist> repository_artist;
        private IRepository<long, Spectacol> repository_spectacol;
        private IRepository<long, Angajat> repository_angajat;
        private IRepository<long, Client> repository_client;
        private IRepository<long, Persoana> repository_persoana;
        private AuthenticationService autentificare;
        private readonly IDictionary <String, IObserver> loggedClients;

        public Service(IRepository<long, Bilet> repo_b,
            IRepository<long, Artist> repo_artist,
            IRepository<long, Spectacol> repo_spectacol,
            IRepository<long, Angajat> repo_angajat,
            IRepository<long, Client> repo_client,
            IRepository<long, Persoana> repo_persoana,
            AuthenticationService autentificare_t
        )
        {
            this.repository_bilet = repo_b;
            this.repository_artist = repo_artist;
            this.repository_spectacol = repo_spectacol;
            this.repository_angajat = repo_angajat;
            this.repository_client = repo_client;
            this.repository_persoana = repo_persoana;
            autentificare = autentificare_t;
            loggedClients=new Dictionary<String, IObserver>();

        }

        
        public List<string> FindArtistiSpectacoleBilete()
        {

            List<string> listaInregistrariBilete = new List<string>();

            foreach (Artist artist in repository_artist.findAll())
            {
                foreach (Spectacol spectacol in repository_spectacol.findAll())
                {
                    if (spectacol.IdArtist == artist.IdPersoana)
                    {
                        int nr_bilete_date = 0;
                        foreach (Bilet bilet in repository_bilet.findAll())
                        {
                            if (bilet.IdSpectacol == spectacol.IdSpectacol)
                            {
                                nr_bilete_date += bilet.NrLocuri;
                            }
                        }

                        listaInregistrariBilete.Add(artist.Nume + ' ' + artist.Prenume + ','
                                                    + spectacol.DataIncepere + ',' + spectacol.Locatie + ',' +
                                                    spectacol.NumarLocuri + ',' + nr_bilete_date);
                    }
                }
            }


            return listaInregistrariBilete;
        }

        public List<string> FindArtistiSpectacoleBileteDupaData(DateTime date)
        {

            List<string> listaInregistrariBilete = new List<string>();

            foreach (Artist artist in repository_artist.findAll())
            {
                foreach (Spectacol spectacol in repository_spectacol.findAll())
                {
                    if (spectacol.DataIncepere.Day == date.Day && spectacol.DataIncepere.Month == date.Month &&
                        spectacol.DataIncepere.Year == date.Year)
                        if (spectacol.IdArtist == artist.IdPersoana)
                        {
                            int nr_bilete_date = 0;
                            foreach (Bilet bilet in repository_bilet.findAll())
                            {
                                if (bilet.IdSpectacol == spectacol.IdSpectacol)
                                {
                                    nr_bilete_date += bilet.NrLocuri;
                                }
                            }

                            listaInregistrariBilete.Add(artist.Nume + ' ' + artist.Prenume + ','
                                                        + spectacol.DataIncepere + ',' + spectacol.Locatie + ',' +
                                                        spectacol.NumarLocuri + ',' + nr_bilete_date);
                        }
                }
            }

            return listaInregistrariBilete;
        }


        public Spectacol FindArtistiSpectacoleBileteIndex(int index)
        {

            List<string> listaInregistrariBilete = new List<string>();
            int i = 0;
            foreach (Artist artist in repository_artist.findAll())
            {
                foreach (Spectacol spectacol in repository_spectacol.findAll())
                {
                    if (spectacol.IdArtist == artist.IdPersoana)
                    {
                        if (i == index)
                            return spectacol;
                        i += 1;
                    }
                }
            }

            return null;
        }


        public Spectacol FindArtistiSpectacoleBileteDupaDataIndex(DateTime date, int index)
        {

            List<string> listaInregistrariBilete = new List<string>();
            int i = 0;
            foreach (Artist artist in repository_artist.findAll())
            {
                foreach (Spectacol spectacol in repository_spectacol.findAll())
                {
                    if (spectacol.DataIncepere.Day == date.Day && spectacol.DataIncepere.Month == date.Month &&
                        spectacol.DataIncepere.Year == date.Year)
                        if (spectacol.IdArtist == artist.IdPersoana)
                        {
                            if (i == index)
                                return spectacol;
                            i += 1;
                        }
                }
            }

            return null;
        }



        public long SavePersReturnIdPers(string nume, string prenume)
        {
            Persoana persoana = new Persoana(nume, prenume);
            repository_persoana.save(persoana);
            List<Persoana> p = (List<Persoana>)repository_persoana.findAll();

            return p[p.Count - 1].IdPersoana;
        }


        public long SaveClientReturnIdClient(string nume, string prenume)
        {
            long id = SavePersReturnIdPers(nume, prenume);
            Client pers = new Client(id, nume, prenume);
            repository_client.save(pers);
            List<Client> p = (List<Client>)repository_client.findAll();

            return p[p.Count - 1].IdPersoana;
        }

        public Bilet SaveBilet(string nume, string prenume, int numarLocuriB, long IDAngajat, long IDSpectacol)
        {
            long idClient = SaveClientReturnIdClient(nume, prenume);
            Bilet bilet = new Bilet(IDAngajat, idClient, IDSpectacol, numarLocuriB);
            repository_bilet.save(bilet);
            foreach (var obs in loggedClients.Values)
            {
                Task.Run(() =>
                    {
                        try
                        {
                            obs.rezervareReceived(bilet);
                        }
                        catch (Exceptions e)
                        {
                            Console.WriteLine("eroare obs srv");
                        }

                    })
                ;

            }
            
            return bilet;
        }

        public Angajat Valid(string pass, string u,IObserver client)
        {

            Angajat userOk = autentificare.Authenticate(u, pass);
            if (userOk!=null){
                if(loggedClients.ContainsKey(userOk.Id.ToString()))
                    throw new Exceptions("User already logged in.");
                
                Console.WriteLine("IDDDD"+userOk.IdPersoana.ToString());
                loggedClients[userOk.IdPersoana.ToString()]= client;
               // notifyFriendsLoggedIn(user);
            }
            // Apelăm metoda authenticate
            return autentificare.Authenticate(u, pass);
            //return new Angajat(122,"nume","prenume");
        }
       /*private void notifyFriendsLoggedOut(Angajat user) {
            IEnumerable<Angajat> friends=userRepository.getFriendsOf(user);
            foreach(User us in friends){
                if (loggedClients.ContainsKey(us.Id))
                {
                    IChatObserver chatClient = loggedClients[us.Id];
                    Task.Run(() =>chatClient.friendLoggedOut(user));
                }
            }
        }*/

        public Angajat logout(Angajat user, IObserver client)
        {
            IObserver localClient=loggedClients[user.IdPersoana.ToString()];
            if (localClient==null)
                throw new Exceptions("User "+user.IdPersoana+" is not logged in.");
            loggedClients.Remove(user.IdPersoana.ToString());
           // notifyFriendsLoggedOut(user);
            return user;
        }
    }
}
